sieve_of_eratosthenes/SieveOfEratosthenes.java

```diff
@@ -1,34 +1,128 @@
-/*
- *  Following implementation of sieve of eratosthenes returns a boolean array which contains true is its index is prime
- *  less than or equal to a number n.
- *  Space complexity : O(n).
- *  Time complexity : O(n).
- */
+import java.util.Arrays;
 
+/**
+ * <p>This class represents a Sieve of Eratosthenes. It exposes an isPrime(int) method.</p>
+ *
+ * <p>When creating a Sieve, you need to specify the max number that you might query.</p>
+ *
+ * <p>Time complexity for creating the Sieve is O(n * n) but for querying is O(1). Space complexity is O(n)</p>
+ *
+ * @see <a href="https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes">Sieve of Eratosthenes</a>
+ */
 public class SieveOfEratosthenes {
 
-    public static boolean[] sieveOfEratosthenes(int n) {
-        int sqrtOfn = (int)Math.sqrt(n) + 1;
-        boolean[] primes = new boolean[n + 1];
-        for (int i = 2; i < n + 1; ++i) {
-            primes[i] = true;
-        }
-        for (int i = 2; i <= sqrtOfn; ++i) {
-            if (primes[i]) {
-                for (int j = 2 * i; j < n + 1; j += i) {
-                    primes[j] = false;
-                }
+    private boolean[] sieve;
+
+    /**
+     * Creates a sieve that can be queried for numbers up to maxNumber.
+     *
+     * @param maxNumber
+     */
+    public SieveOfEratosthenes(final int maxNumber) {
+        initSieve(maxNumber);
+        uncrossAllNumbers();
+        crossObviousNumbers();
+        crossMultiplesOfKnownPrimes();
+    }
+
+    /**
+     * Iterates the sieve and when it finds a prime in it, it'll cross its multiples, since by definition those numbers are not prime.
+     */
+    private void crossMultiplesOfKnownPrimes() {
+        for (int n = 2; nextMultiplePossiblyNotCrossedOf(n) < sieve.length; ++n) {
+            if (isPrime(n)) {
+                crossMultiplesOf(n);
             }
         }
-        return primes;
+    }
+
+    /**
+     * Given prime, crosses all multiples of prime in the sieve
+     *
+     * @param prime
+     */
+    private void crossMultiplesOf(final int prime) {
+        for (int n = nextMultiplePossiblyNotCrossedOf(prime); n < sieve.length; n = nextMultipleOf(prime, n)) {
+            cross(n);
+        }
+    }
+
+    /**
+     * Crosses 0 and 1 which by definition aren't prime.
+     */
+    private void crossObviousNumbers() {
+        cross(0);
+        cross(1);
+    }
+
+    /**
+     * Returns the next multiple of n which is greater than last.
+     *
+     * @param n
+     * @param last
+     * @return
+     */
+    private int nextMultipleOf(final int n, final int last) {
+        return n + last;
+    }
```

Comment:
Having such methods would be an overkill. We can simply use these values wherever needed.

Reply:
@singhpratyush we land in the pros / cons of Clean Code. On one hand you get very expressive code that's really simple to understand and is really simple to test, but you have a lot of methods. The key aspect for me is to consider: is the code by itself clear, or having a method with a good name helps it make it even more clear. How many people will see the line i += j and immediately realize that are jumping through all the multiples of j? Instead, when you have a method that is called nextMultipleOf() everyone will understand what's the purpose of that, even if they don't understand why n + last is the next multiple of n greater than last. That's the point of levels of indirection: you can hide some implementations details (in this case a Mathematical fact) behind a named function, which will make it easier to understand. Only if you want to understand the details then you need to read the n + last code.

Reply:
@singhpratyush I forget to mention, if you think having a lot of method is overkill because of performance concerns, it turns out that the compiler will surely inline methods like this so you have the advantage of expressive code without the penalty on performance. But in general I think readability trumps performance.